### DISCUSSION TITLE
which python file is responsible for showing meson's help

### DISCUSSION THREAD

**Discussion by MadadiSaeed** [mesonbuild/meson#9966] [Q&A] [answered]
as you can see from the meson's help ouput there are two additional & useless sections that the first one can be shorten to `meson [-h] [Commands] ...` and the second one can be completely removed from the output. how can i edit this output?
**Note**: I am talking about this section of output. `{setup,configure,dist,install,introspect,init,test,wrap,subprojects,help,rewrite,compile,devenv}`
![Screenshot from 2022-02-08 02-17-02](https://user-images.githubusercontent.com/18693428/152885846-9712ddcb-1ceb-459d-9b4f-0e64185bf5df.png)

Second thing that i want to ask is, should i open a new issue about showing help about meson sub-commands?

**Comment by eli-schwartz** [comment, marked as answer]
This is implemented in mesonbuild/mesonmain.py as an https://docs.python.org/3/library/argparse.html client. So the actual formatting of the help text is largely up to whatever the python developers decided argparse should do.

**Reply by dcbaker** [reply]
You can set flags to `argparse.ArgumentParser` to control that. But, it seems like a matter of personal taste whether `meson [command] or `meson {setup, configure, ...}` is better, and *someone* is going to be unhappy either way

**Reply by MadadiSaeed** [reply]
Two things that is wrong about this format:
First: it is repetitive. when you have dedicated section for commands that even have their descriptions, i don't think it is good idea to add same extra text just for matter of taste. 
Second: the logic is wrong and confusing. this format has three sections ( `[-h]`, `{setup, configure, ...}`, `...` ). first section (`[-h]`) is optional and it is about showing help of something, but what is it? is it about showing meson's help or meson sub-command's help? second section i think it is required but when you use `[-h]` before that, it should show sub-command's help but it doesn't. it prints same output from before. third section is `...`. but what is this? is it a folder name or project name or file name? what should i put here? maybe it depends on sub-command.
As a user of this professional tool, I think it should has a very good command line interface.

**Reply by dcbaker** [reply]
What you're running into is that meson used to have separate commands `meson`, `meson-configure`, etc. Then we merged into a git-like interface with verbage. When that happened `meson` became `meson setup`, but for backwards compatibility we had to make `meson` act mostly like `meson setup`.
